File: composeApp/src/commonMain/kotlin/com/aus/notelikeus/data/remote/AttachmentObjectKey.kt
package com.aus.notelikeus.data.remote

private val NOTE_ID_PATTERN = Regex("^[A-Za-z0-9._-]{1,128}$")
private val ATTACHMENT_ID_PATTERN = Regex("^[A-Za-z0-9._-]{1,128}$")

object AttachmentObjectKey {
    fun build(ownerId: String, noteId: String, attachmentId: String): String {
        val trimmedNoteId = noteId.trim()
        val trimmedAttachmentId = attachmentId.trim()
        require(NOTE_ID_PATTERN.matches(trimmedNoteId) && ATTACHMENT_ID_PATTERN.matches(trimmedAttachmentId)) {
            "invalid attachment path segment"
        }
        return "owners/$ownerId/notes/$trimmedNoteId/$trimmedAttachmentId"
    }

    fun workerPath(noteId: String, attachmentId: String): String =
        "/v1/attachments/${encode(noteId.trim())}/${encode(attachmentId.trim())}"

    fun isForOwner(objectKey: String, ownerId: String): Boolean {
        val prefix = "owners/$ownerId/notes/"
        if (!objectKey.startsWith(prefix)) return false
        val rest = objectKey.removePrefix(prefix)
        val slash = rest.indexOf('/')
        if (slash <= 0 || slash == rest.lastIndex) return false
        val noteId = rest.substring(0, slash)
        val attachmentId = rest.substring(slash + 1)
        return NOTE_ID_PATTERN.matches(noteId) && ATTACHMENT_ID_PATTERN.matches(attachmentId)
    }

    private fun encode(value: String): String = buildString(value.length) {
        for (char in value) {
            when {
                char.isLetterOrDigit() || char == '.' || char == '_' || char == '-' -> append(char)
                else -> {
                    val hex = char.code.toString(16).uppercase()
                    append('%')
                    append(hex.padStart(2, '0'))
                }
            }
        }
    }
}

File: web/src/lib/attachments/attachmentObjectKey.ts
const NOTE_ID_PATTERN = /^[A-Za-z0-9._-]{1,128}$/;
const ATTACHMENT_ID_PATTERN = /^[A-Za-z0-9._-]{1,128}$/;

export function buildAttachmentObjectKey(
  ownerId: string,
  noteId: string,
  attachmentId: string,
): string {
  const trimmedNoteId = noteId.trim();
  const trimmedAttachmentId = attachmentId.trim();
  if (!NOTE_ID_PATTERN.test(trimmedNoteId) || !ATTACHMENT_ID_PATTERN.test(trimmedAttachmentId)) {
    throw new Error('invalid attachment path segment');
  }
  return `owners/${ownerId}/notes/${trimmedNoteId}/${trimmedAttachmentId}`;
}

export function isAttachmentObjectKeyForOwner(objectKey: string, ownerId: string): boolean {
  const prefix = `owners/${ownerId}/notes/`;
  if (!objectKey.startsWith(prefix)) return false;
  const rest = objectKey.slice(prefix.length);
  const slash = rest.indexOf('/');
  if (slash <= 0 || slash === rest.length - 1) return false;
  const noteId = rest.slice(0, slash);
  const attachmentId = rest.slice(slash + 1);
  return NOTE_ID_PATTERN.test(noteId) && ATTACHMENT_ID_PATTERN.test(attachmentId);
}

export function attachmentWorkerPath(noteId: string, attachmentId: string): string {
  return `/v1/attachments/${encodeURIComponent(noteId.trim())}/${encodeURIComponent(attachmentId.trim())}`;
}
